substrate/client/network/sync/src/strategy.rs

```diff
@@ -92,15 +94,68 @@ pub enum SyncingAction<B: BlockT> {
 		number: NumberFor<B>,
 		justifications: Justifications,
 	},
-	/// Syncing strategy has finished.
+	/// Strategy finished. Nothing to do, this is handled by `SyncingStrategy`.
 	Finished,
 }
 
+impl<B: BlockT> SyncingAction<B> {
+	fn is_finished(&self) -> bool {
+		matches!(self, SyncingAction::Finished)
+	}
+}
+
+impl<B: BlockT> From<WarpSyncAction<B>> for SyncingAction<B> {
+	fn from(action: WarpSyncAction<B>) -> Self {
+		match action {
+			WarpSyncAction::SendWarpProofRequest { peer_id, request } =>
+				SyncingAction::SendWarpProofRequest { peer_id, request },
+			WarpSyncAction::SendBlockRequest { peer_id, request } =>
+				SyncingAction::SendBlockRequest { peer_id, request },
+			WarpSyncAction::DropPeer(bad_peer) => SyncingAction::DropPeer(bad_peer),
+			WarpSyncAction::Finished => SyncingAction::Finished,
+		}
+	}
+}
+
+impl<B: BlockT> From<StateStrategyAction<B>> for SyncingAction<B> {
+	fn from(action: StateStrategyAction<B>) -> Self {
+		match action {
+			StateStrategyAction::SendStateRequest { peer_id, request } =>
+				SyncingAction::SendStateRequest { peer_id, request },
+			StateStrategyAction::DropPeer(bad_peer) => SyncingAction::DropPeer(bad_peer),
+			StateStrategyAction::ImportBlocks { origin, blocks } =>
+				SyncingAction::ImportBlocks { origin, blocks },
+			StateStrategyAction::Finished => SyncingAction::Finished,
+		}
+	}
+}
+
+impl<B: BlockT> From<ChainSyncAction<B>> for SyncingAction<B> {
+	fn from(action: ChainSyncAction<B>) -> Self {
+		match action {
+			ChainSyncAction::SendBlockRequest { peer_id, request } =>
+				SyncingAction::SendBlockRequest { peer_id, request },
+			ChainSyncAction::SendStateRequest { peer_id, request } =>
+				SyncingAction::SendStateRequest { peer_id, request },
+			ChainSyncAction::CancelRequest { peer_id } => SyncingAction::CancelRequest { peer_id },
+			ChainSyncAction::DropPeer(bad_peer) => SyncingAction::DropPeer(bad_peer),
+			ChainSyncAction::ImportBlocks { origin, blocks } =>
+				SyncingAction::ImportBlocks { origin, blocks },
+			ChainSyncAction::ImportJustifications { peer_id, hash, number, justifications } =>
+				SyncingAction::ImportJustifications { peer_id, hash, number, justifications },
+		}
+	}
+}
+
 /// Proxy to specific syncing strategies.
-pub enum SyncingStrategy<B: BlockT, Client> {
-	WarpSyncStrategy(WarpSync<B, Client>),
-	StateSyncStrategy(StateStrategy<B>),
-	ChainSyncStrategy(ChainSync<B, Client>),
+pub struct SyncingStrategy<B: BlockT, Client> {
+	config: SyncingConfig,
+	client: Arc<Client>,
+	warp: Option<WarpSync<B, Client>>,
+	state: Option<StateStrategy<B>>,
+	chain_sync: Option<ChainSync<B, Client>>,
+	peer_pool: Arc<Mutex<PeerPool>>,
+	peer_best_blocks: HashMap<PeerId, (B::Hash, NumberFor<B>)>,
```

Comment:
It's needed to seed the peers when switching between the strategies. Otherwise, for example, the state strategy won't be aware of the best hash/number (and won't know what peer to request a state from) until the block announcement is received from a peer.